### PR TITLE
Fixes in Diet and tests

### PR DESCRIPTION
**Pull request by larsrh**
I'm not 100% convinced that my changes are accurate. But at least #147 and #149 seem to be fixed.

### CODE REVIEW

**Comment by johnynek** [comment]
@larsrh by the way, if you want to merge this, I am willing to say +1 and I can follow up by adding a bunch of the tests that I have suggested if you don't have time to get to it. Collections are some of the most testable examples of code, but also somewhat subtle (as we have seen here with a bug), so I'd love to shoot for 100% coverage.

**Comment by larsrh** [comment]
Your comments are all good (I knew it would be worthwhile to ask you for a review!).

**Comment by larsrh** [comment]
@johnynek I have implemented most of your suggestions. I can live without coverage of a function that's just an alias, though.

OTOH, I have deprecated `map`, because upon closer inspection, its semantics is entirely unclear. The desired list homomorphism property can't hold for arbitrary functions. For example, let's say you have a `Range(1,2)` and you map it with `_*2`. That would give you `Range(2,4)`. But `List(2,3,4)` is not equal to `List(2,4)` (which the law would require). The current implementation seems broken and not worth fixing.

File: core/src/main/scala/cats/collections/Diet.scala [hunk 1]
```diff
@@ -141,21 +144,27 @@ sealed abstract class Diet[A] {
...
   }
 
+  /**
+    * remove a range from Diet
+    */
+  def removeRange(range: Range[A])(implicit enum: Discrete[A], order: Order[A]): Diet[A] =
+    if (order.lteqv(range.start, range.end))
+      removeRangeIncreasing(range)
+    else
+      removeRangeIncreasing(range.reverse)
```
Comment (johnynek) [inline core/src/main/scala/cats/collections/Diet.scala]: This is surprising to me. I would have thought a range with end < start would be an empty Range. Is this not the case?

File: tests/src/test/scala/cats/collections/DietSpec.scala [hunk 2]
```diff
@@ -1,375 +1,231 @@
...
-  test("add disjoint range"){
+  test("shown empty"){
     val diet = Diet.empty[Int]
 
-    val result = diet.addRange(Range(0, 100))
-
-    val other = result.intervals.map(l => l.generate)
-
-    other should matchTo(List(Range(0, 100).toList))
+    diet.show should be ("{ }")
```
Comment (johnynek) [inline tests/src/test/scala/cats/collections/DietSpec.scala]: Personally I’d like the string Diet to appear somewhere here

File: tests/src/test/scala/cats/collections/DietSpec.scala [hunk 3]
```diff
@@ -1,375 +1,231 @@
...
+  test("merge")(forAll { (rs1: Ranges, rs2: Ranges) =>
+    val diet1 = rs1.toDiet
+    val diet2 = rs2.toDiet
+    (diet1 ++ diet2).toList should be((rs1.toSet ++ rs2.toSet).toList.sorted)
+  })
 
-  test("intersect with another diet"){
-    val diet = Diet.empty[Int] + Range(20, 30)
+  test("intersection range")(forAll { (rs: Ranges, m: Int, n: Int) =>
+    whenever(m > n) {
```
Comment (johnynek) [inline tests/src/test/scala/cats/collections/DietSpec.scala]: I think this can be >=

File: core/src/main/scala/cats/collections/Diet.scala (hunk 1 above)
Comment (johnynek) [reply core/src/main/scala/cats/collections/Diet.scala]: I guess this is a property of Range I didn’t know. I guess there are some reasons for it (maybe it makes map on Range more useful, or maybe just more confusing, not sure which).

File: tests/src/test/scala/cats/collections/DietSpec.scala (hunk 2 above)
Comment (larsrh) [reply tests/src/test/scala/cats/collections/DietSpec.scala]: This is a separate discussion, I think. I'll open a ticket.

File: core/src/main/scala/cats/collections/Diet.scala (hunk 1 above)
Comment (larsrh) [reply core/src/main/scala/cats/collections/Diet.scala]: The `Range` semantics admits `start > end` and then walks backwards (see the sources). The tests produce such ranges and apparently it works everywhere.

File: tests/src/test/scala/cats/collections/DietSpec.scala [hunk 4]
```diff
@@ -1,375 +1,231 @@
...
+      // there may be structural changes, so fall back to list comparison
+      d.add(elem).toList should be(d.toList)
+    )
+  })
 
-    diet.toList() should matchTo(List(9, 10, 12, 13, 14, 20, 21, 23, 24))
+  def invariant[A](d: Diet[A])(implicit order: Order[A], enum: Discrete[A]): Boolean = d match {
+    case Diet.DietNode(rng, left, right) =>
+      order.lteqv(rng.start, rng.end) &&
+        left.toList.forall(order.lt(_, rng.start)) &&
```
[the remark this answers is not in the document]
Comment (larsrh) [reply tests/src/test/scala/cats/collections/DietSpec.scala]: Right, it could be tighter. I suppose I would need to compare the predecessor of rng.start.

**Pull request merged**